learning-another-language/objects.md

```diff
@@ -0,0 +1,412 @@
+
+### !end-callout
+
+### Square Bracket Notation
+
+We can also use square bracket notation. Consider again this object:
+
+<!-- prettier-ignore-start -->
+```js
+const testObj = {
+  someNum: 5,
+  someStr: 'this is a test string',
+  someNestedObj: {
+    someOtherNum: 4,
+  }
+};
+```
+<!-- prettier-ignore-end -->
+
+The following code gives us `'this is a test string'`.
+
+<!-- prettier-ignore-start -->
+```js
+testObj['someStr'];
+```
+<!-- prettier-ignore-end -->
+
+Note that square bracket notation allows us to pass in a _variable_ for our key. This is because we can put in expressions in our square brackets. [Observe this code](https://replit.com/@adacore/Objects-Demo#index.js):
+
+<!-- prettier-ignore-start -->
+```js
+const keyToGetSomeValue = 'someStr';
+
+testObj[keyToGetSomeValue];
+```
+<!-- prettier-ignore-end -->
+
+We declared a `const` variable named `keyToGetSomeValue`, and assigned it to `'someStr'`. When we use `keyToGetSomeValue` in square brackets while accessing `testObj`, this variable evaluates to `someStr`.
+
+### Nested Objects
+
+We can nest objects inside other objects.
+
+<!-- prettier-ignore-start -->
+```js
+const testObj = {
+  someNestedObj: {
+    someOtherNum: 4,
+  }
+};
+```
+<!-- prettier-ignore-end -->
+
+We can infinitely nest objects!
+
+In order to access nested objects, we continue to _chain_ our syntax.
+
+To get `4` from accessing `testObj`, we can use [any of the following syntax](https://replit.com/@adacore/Objects-Demo#index.js):
+
+<!-- prettier-ignore-start -->
+```js
+testObj.someNestedObj.someOtherNum;
+
+testObj['someNestedObj']['someOtherNum'];
+
+testObj.someNestedObj['someOtherNum'];
+
+testObj['someNestedObj'].someOtherNum;
+```
+<!-- prettier-ignore-end -->
+
+We evaluate these lines from left to right.
+1. First, `testObj.someNestedObj` or `testObj['someNestedObj']` accesses the inner object
+1. Then, we use the syntax with `.someOtherNum` or `['someOtherNum']` on the inner object to access its own key-value pairs.
+
+Note that we _can_ mix syntaxes in one line. This might be useful in _very_ specific cases, though the rest of the time we should strive for consistency.
+
+## Functions Are Objects
+
+Functions are objects, too!
+
+Just like the object literals we've been working with, functions _also_ have key-value pairs, even though we don't normally see them.
+
+Here are some example properties that all functions have:
+
+- `name`, which is the name of the function
+- `length`, which is the number of parameters the function expects
+
+We won't normally use these properties, but they illustrate a powerful point: we can access these properties using dot notation or square bracket notation, just like any other object.
+
+[Observe this code](https://replit.com/@adacore/Objects-Demo#index.js):
+
+<!-- prettier-ignore-start -->
+```js
+const sayHelloWorld = function() {
```

Comment:
From reading I've done in the past, technically this is an anonymous function, and it's only through interpreter convention that it looks for the assignment to a variable to try an associate a name with the function, largely to aid with debugging. We can see in other situations that (passing a new function expression as an expression to another function) that the name property will not be populated. Consider:
```js
const pn = function(fn) { console.log(fn.name); };
pn(pn);  // prints "pn". Not standard, but everything does it 🤷 
pn(function(){});  // prints ""
```

Reply:
Turning this into an issue to think about it later!